### PR TITLE
PRP: GeoServer Unauthenticated Remote Code Execution 

### PR DESCRIPTION
**Pull request by grandsilva**

### CODE REVIEW

**Comment by lokiuox** [comment]
Hey @grandsilva, I've checked your commits and they look ok, but the `detect_whenVulnerable_returnsVulnerability` test fails. I think it's because of the new version detection step, so please make sure to account for that in the tests too.

File: community/detectors/geoserver_cve_2024_36401/src/main/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetector.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("GeoserverCve202436401VulnDetector starts detecting.");
+
+    return DetectionReportList.newBuilder()
+        .addAllDetectionReports(
+            matchedServices.stream()
+                .filter(this::isGeoserverInstance)
```
Comment (lokiuox) [inline community/detectors/geoserver_cve_2024_36401/src/main/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetector.java:101]:
```suggestion
            matchedServices.stream()
                .filter(NetworkServiceUtils::isWebService)
                .filter(this::isGeoserverInstance)
```

Just to be safe, let's filter for web services as the first step.

File: community/detectors/geoserver_cve_2024_36401/src/test/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetectorTest.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  @Test
+  public void detect_ifNotVulnerable_doesNotReportVuln() throws IOException {
+    startMockWebServer();
+    mockCallbackServer.enqueue(PayloadTestHelper.generateMockUnsuccessfulCallbackResponse());
+    DetectionReportList detectionReports =
+        detector.detect(targetInfo, ImmutableList.of(targetNetworkService));
+    assertThat(detectionReports.getDetectionReportsList()).isEmpty();
+    Truth.assertThat(mockTargetService.getRequestCount()).isEqualTo(1);
```
Comment (lokiuox) [inline community/detectors/geoserver_cve_2024_36401/src/test/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetectorTest.java:152]:
```suggestion
    Truth.assertThat(mockTargetService.getRequestCount()).isEqualTo(2);
```
You missed this in the last commit

**Review by lokiuox — CHANGES_REQUESTED** [review]
Hey @grandsilva, the plugin looks good. I've left comments on two small things to fix, then it's good to go. 

Also, please make sure to run the google-java-format tool before pushing new changes (the command-line version is preferred).

File: community/detectors/geoserver_cve_2024_36401/src/main/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetector.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+import com.google.tsunami.common.net.http.HttpStatus;
+import com.google.tsunami.common.time.UtcClock;
+import com.google.tsunami.plugin.PluginType;
+import com.google.tsunami.plugin.VulnDetector;
+import com.google.tsunami.plugin.annotations.ForWebService;
+import com.google.tsunami.plugin.annotations.PluginInfo;
+import com.google.tsunami.plugin.payload.NotImplementedException;
+import com.google.tsunami.plugins.detectors.rce.Annotations.OobSleepDuration;
+import com.google.tsunami.plugin.payload.Payload;
+import com.google.tsunami.plugin.payload.PayloadGenerator;
```
Comment (lokiuox) [inline community/detectors/geoserver_cve_2024_36401/src/main/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetector.java:40]:
```suggestion
import com.google.tsunami.plugin.payload.Payload;
import com.google.tsunami.plugin.payload.PayloadGenerator;
import com.google.tsunami.plugins.detectors.rce.Annotations.OobSleepDuration;
```
Reorder imports

File: community/detectors/geoserver_cve_2024_36401/src/main/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetector.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.tsunami.common.data.NetworkServiceUtils.buildWebApplicationRootUrl;
+import static com.google.tsunami.common.net.http.HttpRequest.get;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.flogger.GoogleLogger;
+import com.google.common.util.concurrent.Uninterruptibles;
+import com.google.protobuf.util.Timestamps;
+import com.google.tsunami.common.net.http.HttpClient;
```
Comment (lokiuox) [inline community/detectors/geoserver_cve_2024_36401/src/main/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetector.java:29]:
```suggestion
import com.google.tsunami.common.data.NetworkServiceUtils;
import com.google.tsunami.common.net.http.HttpClient;
```
Add import for NetworkServiceUtils

File: community/detectors/geoserver_cve_2024_36401/src/main/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetector.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                .setSeverity(Severity.CRITICAL)
+                .setTitle("Geoserver RCE (CVE-2024-36401)")
+                .setDescription(
+                    "This detector checks for Geoserver RCE (CVE-2024-36401). "
+                        + "Multiple OGC request parameters allow Remote Code Execution (RCE) "
+                        + "by unauthenticated users through specially crafted input against "
+                        + "a default GeoServer installation due to unsafely evaluating property "
+                        + "names as XPath expressions.")
+                .setRecommendation(
+                    "Upgrade Geoserver to a patched version. The vulnerability was fixed in versions 2.23.6, 2.24.4, and 2.25.2.")
```
Comment (lokiuox) [inline community/detectors/geoserver_cve_2024_36401/src/main/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetector.java:192]:
```suggestion
                    "Upgrade Geoserver to a patched version. The vulnerability was fixed in"
                        + " versions 2.23.6, 2.24.4, and 2.25.2.")
```
Split overlong line

File: community/detectors/geoserver_cve_2024_36401/src/test/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetectorTest.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+import com.google.inject.testing.fieldbinder.Bind;
+import com.google.inject.testing.fieldbinder.BoundFieldModule;
+import com.google.inject.util.Modules;
+import com.google.protobuf.util.Timestamps;
+import com.google.tsunami.common.net.http.HttpClientModule;
+import com.google.tsunami.common.time.testing.FakeUtcClock;
+import com.google.tsunami.common.time.testing.FakeUtcClockModule;
+import com.google.tsunami.plugin.payload.testing.FakePayloadGeneratorModule;
+import com.google.tsunami.plugins.detectors.rce.Annotations.OobSleepDuration;
+import com.google.tsunami.plugin.payload.testing.PayloadTestHelper;
```
Comment (lokiuox) [inline community/detectors/geoserver_cve_2024_36401/src/test/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetectorTest.java:34]:
```suggestion
import com.google.tsunami.plugin.payload.testing.PayloadTestHelper;
import com.google.tsunami.plugins.detectors.rce.Annotations.OobSleepDuration;
```
Reorder imports

File: community/detectors/geoserver_cve_2024_36401/src/test/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetectorTest.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                        .setSeverity(Severity.CRITICAL)
+                        .setTitle("Geoserver RCE (CVE-2024-36401)")
+                        .setDescription(
+                            "This detector checks for Geoserver RCE (CVE-2024-36401). "
+                                + "Multiple OGC request parameters allow Remote Code Execution (RCE) "
+                                + "by unauthenticated users through specially crafted input against "
+                                + "a default GeoServer installation due to unsafely evaluating property "
+                                + "names as XPath expressions.")
+                        .setRecommendation(
+                            "Upgrade Geoserver to a patched version. The vulnerability was fixed in versions 2.23.6, 2.24.4, and 2.25.2.")
```
Comment (lokiuox) [inline community/detectors/geoserver_cve_2024_36401/src/test/java/com/google/tsunami/plugins/detectors/rce/GeoserverCve202436401VulnDetectorTest.java:135]:
```suggestion
                        .setDescription(
                            "This detector checks for Geoserver RCE (CVE-2024-36401). Multiple OGC"
                                + " request parameters allow Remote Code Execution (RCE) by"
                                + " unauthenticated users through specially crafted input against a"
                                + " default GeoServer installation due to unsafely evaluating"
                                + " property names as XPath expressions.")
                        .setRecommendation(
                            "Upgrade Geoserver to a patched version. The vulnerability was fixed in"
                                + " versions 2.23.6, 2.24.4, and 2.25.2.")
```
Split overlong strings

**Review by lokiuox — CHANGES_REQUESTED** [review]
Hey @grandsilva, I cannot push to your branch as I do not have write access to it, but there were just a few changes so I put them here as review suggestions so you can auto-apply them.

**Comment by lokiuox** [comment]
LGTM - Approved
@maoning we can merge this and https://github.com/google/security-testbeds/pull/79

Reviewer: Savio, Doyensec
Plugin: GeoServer Unauthenticated RCE - CVE-2024-36401
Feedback: The overall quality of the contribution is very good. The plugin only lacked a fingerprinting step, but the implementation quality was high; the text description could have been improved a bit, too. The testbed was a bit too complex and could be simplified. The contributor was very responsive.

**Review by tooryx — APPROVED** [review]

**Comment by tooryx** [comment]
Hi @grandsilva,

This PR is currently in internal review and should be merge in a few days.

~tooryx

**Pull request merged**